### PR TITLE
feat: RFC0025 didcomm transports tests

### PR DESCRIPTION
**Pull request by TimoGlastra**
Adds testing for RFC 0025 didcomm transports.

- Implement the `/agent/command/agent/start` command for ACA-Py to be able to restart the agent with different parameters
- Allow to configure the inbound and outbound transports an agent should use in the `/agent/command/agent/start` command.
- Supports HTTP and WS transport
- Both HTTP and WS are ran for AFGO and ACA-Py backchannels
- WS transport is disabled for the acapy-main <-> afgo-interop runsets as I couldn't get websockets to work between ACA-Py and AFGO
- Extended mediation tests. It now runs a test with two agents connecting through a mediator that don't have an endpoint themselves. 
- Added tags to filter out some tests for ACA-Py as it can't run without inbound transports (`@Transport_NoInbound`) and also can't run without http outbound transport (`@Transport_NoHttpOutbound`) because it using http outbound transport to send webhook events

To easily run the tests:

**AFGO**
```sh
LEDGER_URL_CONFIG=http://test.bcovrin.vonx.io TAILS_SERVER_URL_CONFIG=https://tails.vonx.io ./manage run -d afgo-master -t @RFC0025,@RFC0211 -t ~@wip -v 20 -t ~@T003-RFC0211
```

**ACA-Py**
```sh
LEDGER_URL_CONFIG=http://test.bcovrin.vonx.io TAILS_SERVER_URL_CONFIG=https://tails.vonx.io ./manage run -d acapy-main -t @RFC0025,@RFC0211 -t ~@wip -t ~@Transport_NoInbound -t ~@Transport_NoHttpOutbound
```

**ACA-Py x AFGO**
```sh
LEDGER_URL_CONFIG=http://test.bcovrin.vonx.io TAILS_SERVER_URL_CONFIG=https://tails.vonx.io ./manage run -d afgo-interop -d acapy-main -t @RFC0025,@RFC0211 -t ~@wip -t ~@Transport_NoHttpOutbound -t ~@Transport_Ws
```

```sh
LEDGER_URL_CONFIG=http://test.bcovrin.vonx.io TAILS_SERVER_URL_CONFIG=https://tails.vonx.io ./manage run -d acapy-main -d afgo-interop -t @RFC0025,@RFC0211 -t ~@wip -t ~@Transport_NoHttpOutbound -t ~@Transport_Ws
```

### CODE REVIEW

File: aries-test-harness/features/0025-didcomm-transports.feature [hunk 1]
```diff
@@ -0,0 +1,66 @@
+@RFC0025 @AIP20 @UsesCustomParameters @DIDExchangeConnection
+Feature: RFC 0025 DIDComm Transports
+  In order to communicate with other agents,
+  As an Agent
+  I want to various transport protocol.
```
Comment (regiseloi-idlab) [inline aries-test-harness/features/0025-didcomm-transports.feature]: @TimoGlastra: something missing in that sentence?

**Review by regiseloi-idlab — COMMENTED** [review]
This looks good @TimoGlastra. Thank you for the walkthrough of @T001-RFC0025. Mixing transport options and using mediation obviously adds some complexity, but this approach gives us what we were looking for to enhance AATH in that area.

File: aries-test-harness/features/0025-didcomm-transports.feature (hunk 1 above)
Comment (TimoGlastra) [reply aries-test-harness/features/0025-didcomm-transports.feature]: Thanks, fixed!

**Review by nodlesh — CHANGES_REQUESTED** [review]
Still running some tests, but this looks fantastic. Thanks for this effort. 
One small thing. Can we get `@critical` `@normal` `@minor` tags on all the scenarios to classify the test priority? These tags work with allure for reporting.

Can we also use the `@AcceptanceTest` tag for scenarios that are directly derived from what is stated in the RFC? Also use `@ExceptionTest`, `@NegativeTest` etc, if it makes sense. 

Thanks!

**Comment by nodlesh** [comment]
I can' get all the tests working in the acapy-main scenario you give above. They all don't pass consistently, but after cleaning up some containers, the only one consistently failing is this one. Is this expected?
```
  @T002-RFC0025 @Transport_Http @Transport_Ws
  Scenario Outline: Fail creating a connection between two agents that have mismatching transports configured -- @1.1 Both agents having no overlap in inbound and outbound transports  # features/0025-didcomm-transports.feature:66
    Given we have "2" agents                                                                                                                                                            # features/steps/0160-connection.py:22 0.000s | name | role      |
      | name | role      |
      | Acme | responder |
      | Bob  | requester |
    And "Acme" is running with parameters "{"inbound_transports": ["ws"], "outbound_transports": ["http"] }"                                                                            # features/steps/0044-mime-types.py:12 3.622s
    And "Bob" is running with parameters "{"inbound_transports": ["ws"], "outbound_transports": ["http"] }"                                                                             # features/steps/0044-mime-types.py:12 4.586s
    When "Acme" sends an explicit invitation to "Bob"                                                                                                                                   # features/steps/0023-did-exchange.py:40 0.060s
    And "Bob" receives the invitation                                                                                                                                                   # features/steps/0023-did-exchange.py:204 0.040s
    When "Bob" sends the request to "Acme"                                                                                                                                              # features/steps/0023-did-exchange.py:241 0.068s
    Then "Acme" does not receive the request                                                                                                                                            # features/steps/0023-did-exchange.py:254 2.565s
      Traceback (most recent call last):
        File "/usr/local/lib/python3.7/site-packages/behave/model.py", line 1329, in run
          match.run(runner.context)
        File "/usr/local/lib/python3.7/site-packages/behave/matchers.py", line 98, in run
          self.func(context, *args, **kwargs)
        File "features/steps/0023-did-exchange.py", line 281, in step_impl
          assert expected_agent_state(responder_url, "did-exchange", context.connection_id_dict[responder][context.requester_name] , ["invitation-sent"])
      AssertionError
      
      Captured stdout:
      From http://0.0.0.0:9020 Expected state ['invitation-sent', 'N/A'] but received invitation-received , with a response status of 200
```

When tests inconsistently fail, it is usually like this example. 
```
  @T004-RFC0211 @UsesCustomParameters @RFC0025 @Transport_Http @Transport_Ws
  Scenario Outline: Two agents creating a DIDExchange connection using a mediator -- @2.1 Creating a DIDExchange connection using a mediator with overlapping transports  # features/0211-coordinate-mediation.feature:79
    Given we have "2" agents                                                                                                                                              # features/steps/0160-connection.py:22 0.000s
      | name  | role      |
      | Acme  | mediator  |
      | Bob   | recipient |
      | Faber | sender    |
    And "Acme" is running with parameters "{"inbound_transports": ["http", "ws"], "outbound_transports": ["http", "ws"] }"                                                # features/steps/0044-mime-types.py:12 31.472s
      Assertion Failed: resp_status 500 is not 200; 500 Internal Server Error
      
      Server got itself in trouble
      Captured stdout:
      Data table in step contains an unrecognized role 'recipient', must be inviter, invitee, inviteinterceptor, issuer, holder, verifier, prover, requester, responder, mediator, and recipient
      Data table in step contains an unrecognized role 'sender', must be inviter, invitee, inviteinterceptor, issuer, holder, verifier, prover, requester, responder, mediator, and recipient
```

**Comment by nodlesh** [comment]
Most of the afgo - acapy tests work, but in the -d acapy-main -d afgo-interop configuration this happens. Expected?

```
  @T003-RFC0211 @critical @AcceptanceTest
  Scenario: Request mediation with the mediator denying the mediation request  # features/0211-coordinate-mediation.feature:33
    Given we have "2" agents                                                   # features/steps/0160-connection.py:22 0.000s
      | name | role      |
      | Acme | mediator  |
      | Bob  | recipient |
    And "Acme" and "Bob" create a new didexchange connection                   # features/steps/0023-did-exchange.py:21 25.249s
    When "Bob" requests mediation from "Acme"                                  # features/steps/0211-coordinate-mediation.py:10 1.012s
    And "Acme" denies the mediation request from "Bob"                         # features/steps/0211-coordinate-mediation.py:40 0.005s
      Assertion Failed: resp_status 501 is not 200; mediate-deny message not supported by AFGO.
      Captured stdout:
      Data table in step contains an unrecognized role 'recipient', must be inviter, invitee, inviteinterceptor, issuer, holder, verifier, prover, requester, responder, mediator, and recipient

    Then "Acme" has denied the mediation request from "Bob"                    # None

```

**Comment by TimoGlastra** [comment]
Thanks for the feedback Sheldon. I've added the tags as you mention, please let me know if I've marked any of them incorrectly.

> Most of the afgo - acapy tests work, but in the -d acapy-main -d afgo-interop configuration this happens. Expected?

I haven't enabled mediation tests between AFGO and ACA-Py because I couldn't get them to work. However AFGO doesn't support the mediation deny message so it is expected to not work any test that AFGO is involved (the afgo runset includes the `-t ~@T003-RFC0211`)

For the acapy-main tests I also have some failing tests now. Will look into it. Weird as it worked last time I ran them....

**Comment by TimoGlastra** [comment]
@nodlesh updated the PR

**Review by nodlesh — APPROVED** [review]
This looks great. 

I do get one test failing but I think based on your comment this one is expected? 
-a acapy-main -d afgo-interop
```
  @T003-RFC0211 @critical @AcceptanceTest
  Scenario: Request mediation with the mediator denying the mediation request  # features/0211-coordinate-mediation.feature:33
    Given we have "2" agents                                                   # features/steps/0160-connection.py:22 0.000s
      | name | role      |
      | Acme | mediator  |
      | Bob  | recipient |
    And "Acme" and "Bob" create a new didexchange connection                   # features/steps/0023-did-exchange.py:21 25.275s
    When "Bob" requests mediation from "Acme"                                  # features/steps/0211-coordinate-mediation.py:10 1.016s
    And "Acme" denies the mediation request from "Bob"                         # features/steps/0211-coordinate-mediation.py:40 0.021s
      Assertion Failed: resp_status 501 is not 200; mediate-deny message not supported by AFGO.
      Captured stdout:
      Data table in step contains an unrecognized role 'recipient', must be inviter, invitee, inviteinterceptor, issuer, holder, verifier, prover, requester, responder, mediator, and recipient

    Then "Acme" has denied the mediation request from "Bob"                    # None
```

**Comment by TimoGlastra** [comment]
Yes, this is expected

**Pull request merged**